lxd/backup.go

```diff
@@ -0,0 +1,297 @@
+		ContainerOnly:    b.containerOnly,
+		OptimizedStorage: b.optimizedStorage,
+	}
+}
+
+func (b *backup) Id() int {
+	return b.id
+}
+
+func (b *backup) Name() string {
+	return b.name
+}
+
+func (b *backup) CreationDate() time.Time {
+	return b.creationDate
+}
+
+func (b *backup) ExpiryDate() time.Time {
+	return b.expiryDate
+}
+
+func (b *backup) ContainerOnly() bool {
+	return b.containerOnly
+}
+
+func (b *backup) OptimizedStorage() bool {
+	return b.optimizedStorage
+}
+
+func getBackupInfo(r io.Reader) (*backupInfo, error) {
+	var buf bytes.Buffer
+	err := shared.RunCommandWithFds(r, &buf, "unxz", "-")
+	if err != nil {
+		return nil, err
+	}
+
+	result := backupInfo{}
+	hasBinaryFormat := false
+	hasIndexFile := false
+	tr := tar.NewReader(&buf)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		if hdr.Name == "backup/index.yaml" {
+			err = yaml.NewDecoder(tr).Decode(&result)
+			if err != nil {
+				return nil, err
+			}
+
+			hasIndexFile = true
+		}
+
+		if hdr.Name == "backup/container.bin" {
+			hasBinaryFormat = true
+		}
+	}
+
+	if !hasIndexFile {
+		return nil, ErrMissingIndexFile
+	}
+
+	result.HasBinaryFormat = hasBinaryFormat
+	return &result, nil
+}
+
+// fixBackupStoragePool changes the pool information in the backup.yaml. This
+// is done only if the provided pool doesn't exist. In this case, the pool of
+// the default profile will be used.
+func fixBackupStoragePool(c *db.Cluster, b backupInfo) error {
+	// Get the default profile
+	_, profile, err := c.ProfileGet("default")
+	if err != nil {
+		return err
+	}
+
+	_, v, err := shared.GetRootDiskDevice(profile.Devices)
+	if err != nil {
+		return err
+	}
+
+	// Get the default's profile pool
+	_, pool, err := c.StoragePoolGet(v["pool"])
+	if err != nil {
+		return err
+	}
+
+	f := func(path string) error {
+		// Read in the backup.yaml file.
+		backup, err := slurpBackupFile(path)
+		if err != nil {
+			return err
+		}
+
+		// Change the pool in the backup.yaml
+		backup.Pool = pool
+		backup.Container.Devices["root"]["pool"] = "default"
+
+		file, err := os.Create(path)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		data, err := yaml.Marshal(&backup)
+		if err != nil {
+			return err
+		}
+
+		_, err = file.Write(data)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	err = f(shared.VarPath("storage-pools", pool.Name, "containers", b.Name, "backup.yaml"))
+	if err != nil {
+		return err
+	}
+
+	for _, snap := range b.Snapshots {
+		err = f(shared.VarPath("storage-pools", pool.Name, "snapshots", b.Name, snap,
+			"backup.yaml"))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func createBackupIndexFile(container container, backup backup) error {
+	pool, err := container.StoragePool()
+	if err != nil {
+		return err
+	}
+
+	file, err := os.Create(filepath.Join(getBackupMountPoint(pool, backup.Name()), "index.yaml"))
```

Comment:
Ugh, I know this is `go` being `go` but can we please refrain from nesting function calls in function arguments? :) Just writing this sentence makes me dizzy. We have a bunch of these things going on in `lxd/container_lxc.go` too. That should all go away if possible. :)

Reply:
I don't really have a problem with that when a function is designed so that it can't fail, no point in needlessly storing the return in a temporary variable when you're not going to perform any extra checks on it anyway.

Reply:
My point is that it is not pleasant to parse this visually when reviewing or reading code. :)